### PR TITLE
Don't use conditional Bash script style

### PR DESCRIPTION
**Pull request by cardil**
* Addition of ERR stack trace and dump state for OpenShift CI
 * Clean up of code
 * Optimization of catalog source creation

### CODE REVIEW

**Comment by markusthoemmes** [comment]
/assign mgencur 

Please cut this up into smaller chunks. This basically redoes the entirety of our scripts and includes refactors that aren't necessary to do all in one go. As it stands, this is super hard if not impossible to review thoroughly. I asked about the impact in the mail thread exactly for this reason: To avoid having a PR that rewrites the entirety of our bash scripts.

**Comment by cardil** [comment]
/hold

Until rebase on other extracted changes

**Comment by cardil** [comment]
/unhold

Unrelated chunks have been extracted and they are already merged. Please review.

**Comment by markusthoemmes** [comment]
/hold

(to avoid accidental merge)
The failures look genuine.

**Comment by cardil** [comment]
> The failures look genuine.
Yep. I need to fix those.

**Comment by cardil** [comment]
/retest
/unhold

Looks like the issues are now fixed

**Comment by cardil** [comment]
Conformance tests seams to fail predictably with:

```
Reason:"Unschedulable", Message:"0/6 nodes are available: 3 Insufficient cpu, 3 node(s) had taint {node-role.kubernetes.io/master: }, that the pod didn't tolerate."
```

That doesn't seam like a fault of this PR @markusthoemmes Do you have any ideas?

**Comment by cardil** [comment]
I think this is now ready for final review @mgencur @markusthoemmes

File: hack/lib/serverless.bash [hunk 1]
```diff
@@ -145,9 +149,10 @@ function deploy_knativeserving_cr {
   # Deploy the full version of KnativeServing (vs. minimal KnativeServing). The future releases should
   # ensure compatibility with this resource and its spec in the current format.
   # This is a way to test backwards compatibility of the product with the older full-blown configuration.
-  oc apply -n "${SERVING_NAMESPACE}" -f "${rootdir}/test/v1alpha1/resources/operator.knative.dev_v1alpha1_knativeserving_cr.yaml" || return $?
+  oc apply -n "${SERVING_NAMESPACE}" -f "${rootdir}/test/v1alpha1/resources/operator.knative.dev_v1alpha1_knativeserving_cr.yaml"
 
-  timeout 900 '[[ $(oc get knativeserving.operator.knative.dev knative-serving -n $SERVING_NAMESPACE -o=jsonpath="{.status.conditions[?(@.type==\"Ready\")].status}") != True ]]'  || return 7
+  timeout 900 "[[ \$(oc get knativeserving.operator.knative.dev knative-serving \
+-n ${SERVING_NAMESPACE} -o=jsonpath='{.status.conditions[?(@.type==\"Ready\")].status}') != True ]]"
```
Comment (mgencur) [inline hack/lib/serverless.bash:155]: Nit: The indentation here seems wrong.

File: hack/lib/ui.bash [hunk 2]
```diff
@@ -20,17 +20,52 @@ readonly COLOR_CYAN='\e[0;36m'
 readonly COLOR_LIGHT_RED='\e[1;31m'
 readonly COLOR_LIGHT_YELLOW='\e[1;33m'
 
+declare -a ERROR_HANDLERS
+trap 'invoke.error_handlers' ERR
+
+function register.error_handler {
+  local handlerfunc
+  handlerfunc="${1:?Pass an error handler as arg[1]}"
+  logger.debug "Registering a error handler: ${handlerfunc}"
```
Comment (mgencur) [inline hack/lib/ui.bash:29]: Nit: we don't need the `a` here. It should be `an` anyway.

File: test/e2e-tests.sh [hunk 3]
```diff
@@ -10,38 +10,33 @@ if [ -n "$OPENSHIFT_CI" ]; then
...
+downstream_serving_e2e_tests
+downstream_eventing_e2e_tests
 if [[ $TEST_KNATIVE_KAFKA == true ]]; then
- (( !failed )) && ensure_kafka_no_auth || failed=8
- (( !failed )) && downstream_knative_kafka_e2e_tests || failed=9
- (( !failed )) && ensure_kafka_tls_auth || failed=10
- (( !failed )) && downstream_knative_kafka_e2e_tests || failed=11
- (( !failed )) && ensure_kafka_sasl_auth || failed=12
- (( !failed )) && downstream_knative_kafka_e2e_tests || failed=13
+  ensure_kafka_no_auth
```
Comment (mgencur) [inline test/e2e-tests.sh:34]: Perhaps wrong indentation for the lines below?

File: test/e2e-tests.sh [hunk 4]
```diff
@@ -10,38 +10,33 @@ if [ -n "$OPENSHIFT_CI" ]; then
...
+  downstream_knative_kafka_e2e_tests
+  ensure_kafka_tls_auth
+  downstream_knative_kafka_e2e_tests
+  ensure_kafka_sasl_auth
+  downstream_knative_kafka_e2e_tests
 fi
 
-(( failed )) && dump_state
```
Comment (mgencur) [inline test/e2e-tests.sh:44]: Are you dropping dump_state on purpose?

File: test/lib.bash [hunk 5]
```diff
@@ -358,12 +324,18 @@ function teardown {
 
 # == State dumps
 
-function dump_state {
+function dump_state.setup {
```
Comment (mgencur) [inline test/lib.bash:327]: Why are you using this "dot" style? We use underscore in most places and here it's even combined. I'm missing the purpose.

**Review by mgencur — COMMENTED** [review]
Hey, Just a few nits and questions. Otherwise looks great!

File: test/e2e-tests.sh (hunk 4 above)
Comment (cardil) [reply test/e2e-tests.sh:44]: Yeah. `dump_state` is automatically invoked on failures.

File: test/e2e-tests.sh (hunk 3 above)
Comment (cardil) [reply test/e2e-tests.sh:34]: The lines were indented with just one space. Those are inside if statement, so I made it 2 spaces consistent with other places..

File: test/lib.bash (hunk 5 above)
Comment (cardil) [reply test/lib.bash:327]: I use dot to indicate a function grouping inside "psudo" object. The same as `logger.info`, `logger.debug`. Then you know that some functions are related to each other.

This is just a preference, though. I'm not strongly pursuing this style.

**Pull request merged**